File: AWS_GUI_Dev/BAWGUI.Settings/ViewModels/ComboBoxItemsSourceDecorator.vb
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Controls.Primitives
Imports System.Windows.Data


Module ComboBoxItemsSourceDecorator
    Public ReadOnly ItemsSourceProperty As DependencyProperty = DependencyProperty.RegisterAttached("ItemsSource", GetType(IEnumerable), GetType(ComboBoxItemsSourceDecorator), New PropertyMetadata(Nothing, AddressOf ItemsSourcePropertyChanged))

    Sub SetItemsSource(ByVal element As UIElement, ByVal value As IEnumerable)
        element.SetValue(ItemsSourceProperty, value)
    End Sub

    Function GetItemsSource(ByVal element As UIElement) As IEnumerable
        Return CType(element.GetValue(ItemsSourceProperty), IEnumerable)
    End Function

    Private Sub ItemsSourcePropertyChanged(ByVal element As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
        Dim target = TryCast(element, Selector)
        If element Is Nothing Then Return
        Dim originalBinding = BindingOperations.GetBinding(target, Selector.SelectedValueProperty)
        BindingOperations.ClearBinding(target, Selector.SelectedValueProperty)

        Try
            target.ItemsSource = TryCast(e.NewValue, IEnumerable)
        Finally
            If originalBinding IsNot Nothing Then BindingOperations.SetBinding(target, Selector.SelectedValueProperty, originalBinding)
        End Try
    End Sub
End Module

Public Class SelectorBehavior
    Public Shared Function GetKeepSelection(ByVal obj As DependencyObject) As Boolean
        Return CBool(obj.GetValue(KeepSelectionProperty))
    End Function

    Public Shared Sub SetKeepSelection(ByVal obj As DependencyObject, ByVal value As Boolean)
        obj.SetValue(KeepSelectionProperty, value)
    End Sub

    Public Shared ReadOnly KeepSelectionProperty As DependencyProperty = DependencyProperty.RegisterAttached("KeepSelection", GetType(Boolean), GetType(SelectorBehavior), New UIPropertyMetadata(False, New PropertyChangedCallback(AddressOf onKeepSelectionChanged)))

    Private Shared Sub onKeepSelectionChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
        Dim selector = TryCast(d, Selector)
        Dim value = CBool(e.NewValue)

        If value Then
            AddHandler selector.SelectionChanged, AddressOf selector_SelectionChanged
        Else
            RemoveHandler selector.SelectionChanged, AddressOf selector_SelectionChanged
        End If
    End Sub

    Private Shared Sub selector_SelectionChanged(ByVal sender As Object, ByVal e As SelectionChangedEventArgs)
        Dim selector = TryCast(sender, Selector)

        If e.RemovedItems.Count > 0 Then
            Dim deselectedItem = e.RemovedItems(0)

            If selector.SelectedItem Is Nothing Then
                selector.SelectedItem = deselectedItem
                e.Handled = True
            End If
        End If
    End Sub
End Class

File: AWS_GUI_Dev/BAWGUI.Utilities/ComboBoxItemsSourceDecorator.cs
using System.Collections;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Controls.Primitives;
using System.Windows.Data;

namespace BAWGUI.Utilities
{
    public static class ComboBoxItemsSourceDecorator
    {
        public readonly static DependencyProperty ItemsSourceProperty = DependencyProperty.RegisterAttached(
            "ItemsSource", typeof(IEnumerable), typeof(ComboBoxItemsSourceDecorator), new PropertyMetadata(null, ItemsSourcePropertyChanged));

        public static void SetItemsSource(UIElement element, IEnumerable value)
        {
            element.SetValue(ItemsSourceProperty, value);
        }

        public static IEnumerable GetItemsSource(UIElement element)
        {
            return (IEnumerable)element.GetValue(ItemsSourceProperty);
        }

        private static void ItemsSourcePropertyChanged(DependencyObject element, DependencyPropertyChangedEventArgs e)
        {
            var target = element as Selector;
            if (element == null)
                return;
            // Save original binding 
            var originalBinding = BindingOperations.GetBinding(target, Selector.SelectedValueProperty);
            BindingOperations.ClearBinding(target, Selector.SelectedValueProperty);

            try
            {
                target.ItemsSource = e.NewValue as IEnumerable;
            }
            finally
            {
                if (originalBinding != null)
                    BindingOperations.SetBinding(target, Selector.SelectedValueProperty, originalBinding);
            }
        }
    }

    public class SelectorBehavior
    {
        public static bool GetKeepSelection(DependencyObject obj)
        {
            return System.Convert.ToBoolean(obj.GetValue(KeepSelectionProperty));
        }

        public static void SetKeepSelection(DependencyObject obj, bool value)
        {
            obj.SetValue(KeepSelectionProperty, value);
        }

        public static readonly DependencyProperty KeepSelectionProperty = DependencyProperty.RegisterAttached("KeepSelection", typeof(bool), typeof(SelectorBehavior), new UIPropertyMetadata(false, new PropertyChangedCallback(onKeepSelectionChanged)));

        private static void onKeepSelectionChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
            var selector = d as Selector;
            var value = System.Convert.ToBoolean(e.NewValue);

            if (value)
                selector.SelectionChanged += selector_SelectionChanged;
            else
                selector.SelectionChanged -= selector_SelectionChanged;
        }

        private static void selector_SelectionChanged(object sender, SelectionChangedEventArgs e)
        {
            var selector = sender as Selector;

            if (e.RemovedItems.Count > 0)
            {
                var deselectedItem = e.RemovedItems[0];

                if (selector.SelectedItem == null)
                {
                    selector.SelectedItem = deselectedItem;
                    e.Handled = true;
                }
            }
        }
    }

}


